### PR TITLE
chore(deps): incremental update dependencies flagged by renovate [WD-8376]

### PR DESCRIPTION
**Pull request by mas-who**
## Done

- Upgraded wait-on from 5.3.0 -> 7.2.0. Affects cypress testing.
- Upgraded typescript from 4.9.5 -> 5.3.3. Affects build script.
- Upgraded @typescript-eslint/eslint-plugin from 5.62.0 -> 6.19.1. Affects lint-js.
- Upgraded @typescript-eslint/parser from 5.62.0 -> 6.19.1. Affects lint-js.
- Upgraded ts-jest from 27.1.5 -> 29.1.2. Affects tests.
- Upgraded stylelint from 15.11.0 -> 16.2.0. Affects lint-style.
- Upgraded stylelint-config-recommended-scss from 5.0.2 -> 14.0.0. Affects lint-style.
- Upgraded stylelint-order from 5.0.0 -> 6.0.4. Affects lint-style.
- Upgraded stylelint from 2.0.0 -> 5.0.0. Affects lint-style.
- Upgraded prettier from 2.8.8 -> 3.2.4. Affects lint-js.
- Upgraded eslint-config-prettier from 8.10.0 -> 9.1.0. Affects lint-js.
- Upgraded eslint-plugin-prettier from 4.2.1 -> 5.1.3. Affects lint-js.
- Upgraded eslint-plugin-testing-library from 5.11.1 -> 6.2.0. Affects lint-js.
- Upgraded @percy/storybook from 4.3.7 -> 5.0.1. Affects testing.
- Upgraded @testing-library/cypress from 9.0.0 -> 10.0.1. Affects testing.
- Upgraded @testing-library/jest-dom from 5.17.0 -> 6.2.1. Affects testing.
- Upgraded babel-jest from 27.5.1 -> 29.7.0. Affects testing.
- Upgraded cypress from 12.17.4 -> 13.6.3. Affects testing.
- Upgraded jest from 27.5.1 -> 29.7.0. Affects testing.
- Upgraded npm-package-json-lint from 5.4.2 -> 7.1.0. Affects lint.
- Upgraded sass-loader from 10.5.1 -> 14.0.0. Affects build.
- Upgraded @types/jest from 27.5.2 -> 29.5.11. Affects build.
- Upgraded nanoid from 3.3.7 -> 5.0.4. Affects build.
- Added jest-environment-jsdom, this is needed for jest upgrade (jest-environment-jsdom no longer ship with jest since v28)
- use node lts in test github workflow
- updated github action dependencies
- updated Docerfile base node image version to 20
- removed dependency resolution for string-width and strip-ansi and fixed resolution for jackspeak to 2.1.1.

### Storybook

To see rendered examples of all react-components, run:

```shell
yarn start
```

### QA in your project

from `react-components` run:

```shell
yarn build
npm pack
```

Install the resulting tarball in your project with:

```shell
yarn add <path-to-tarball>
```

### QA steps

- Make sure ci completes fine and no tests fails
- Spot checks on storybook demo server

### CODE REVIEW

**Comment by mas-who** [comment]
Not sure if the resolution dependencies `strip-ansi` and `string-width` can be upgraded since then are imported in the `cliui` module (some sub dependency of another package) with commonjs syntax (i.e. require) but latest versions of `strip-ansi` and `string-width`  have es module syntax.

**Review by edlerd — APPROVED** [review]
LGTM

File: Dockerfile [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 # Build stage: Install yarn dependencies
 # ===
-FROM node:18 AS yarn-dependencies
+FROM node:21 AS yarn-dependencies
```
Comment (bartaz) [inline Dockerfile:5]: We used to only do LTS versions of node, which would be `20` in this case.

Comment (mas-who) [reply Dockerfile:5]: ahh okay sure, I just copied this update from the renovate PR. Will update this to 20

**Comment by mas-who** [comment]
> Not sure if the resolution dependencies `strip-ansi` and `string-width` can be upgraded since then are imported in the `cliui` module (some sub dependency of another package) with commonjs syntax (i.e. require) but latest versions of `strip-ansi` and `string-width` have es module syntax.

The issue isn't with strip-ansi or string-width, according to this [issue](https://github.com/isaacs/jackspeak/issues/5) it's actually got to do with interaction between yarn's algorithm and `jackspeak`. The proposed solution is to fix `jackspeak` resolution to 2.1.1 or alternatively upgrade to yarn 3. I chose the first approach as upgrading to yarn 3 is a bit more involved and probably should be included in a separate PR.

File: Dockerfile (hunk 1 above)
Comment (bartaz) [reply Dockerfile]: There is no easy way to tell renovate to only do node LTS "equal" numbers. I think at some point we had it hardcoded in renovate config for couple versions forward, but seems it's gone from the shared config now.

Comment (bartaz) [reply Dockerfile]: I still have it in Vanilla: https://github.com/canonical/vanilla-framework/blob/main/renovate.json#L17-L21

**Review by bartaz — APPROVED** [review]
LGTM

**Pull request merged**